daemon/cmd/watchdogs.go

```diff
@@ -0,0 +1,60 @@
+// SPDX-License-Identifier: Apache-2.0
+// Copyright Authors of Cilium
+
+package cmd
+
+import (
+	"context"
+	"time"
+
+	"github.com/cilium/cilium/pkg/controller"
+	"github.com/cilium/cilium/pkg/datapath/loader"
+	"github.com/cilium/cilium/pkg/endpoint"
+)
+
+const epBPFProgWatchdog = "ep-bpf-prog-watchdog"
+
+// checkEndpointBPFProgs check to ensure endpoints associated BPF programs remain loaded.
+func checkEndpointBPFProgs(d *Daemon, rate time.Duration) {
```

Comment:
This function should be tied to `Daemon` anyway, right? It's already taking `Daemon` as a param.

Reply:
I'm also thinking about Jussi's comment of this code should probably just a Cell on its own that hook into the main daemon Cell. I believe that could allow the config options to be bundled with the Cell and not have to sprawl out all the pieces of this functionality all over the code base.

Reply:
Yep, perhaps it'd be a good exercise to do now. The way the code is now is something we can backport, but indeed would be nice to leave this as separate cell to keep encouraging moving into that style.

Here's roughly how it would look like:

```
type endpointBPFWatchdogConfig struct {
  ...
}
func (def endpointBPFWatchdogConfig) Flags(flags *pflag.Flagger) {
  ...
}

var endpointBPFWatchdogCell = cell.Module(
  "endpoint-bpf-program-watchdog",
  "Periodically checks that all endpoint interfaces have BPF program loaded onto them",

  cell.Config(endpointBPFWatchdogConfig),
  cell.Invoke(registerEndpointBPFWatchdog),
)

type endpointBPFWatchdogParams struct {
  cell.In

  Lifecycle hive.Lifecycle
  Config endpointBPFWatchdogConfig
  DaemonPromise promise.Promise[*Daemon]
  Endpoints endpointmanager.EndpointsLookup
  JobRegistry job.Registry
}

func registerEndpointBPFProgWatchdog(p endpointBPFWatchdogParams) {
  g := p.JobRegistry.NewGroup()
  g.Add(job.Timer(
    "endpoint-bpf-watchdog",
    p.Config.EndpointBPFWatchdogInterval,
    func(ctx context.Context) error {
      return checkEndpointBPFPrograms(ctx, p.EndpointsLookup, p.DaemonPromise)
    }))
  p.Lifecycle.Append(g)
}

func checkEndpointBPFPrograms(ctx context.Context, endpoints endpointmanager.EndpointsLookup, dp promise.Promise[*Daemon]) error {

  eps := endpoints.GetEndpoints()
  for ... { ... }
  if loaded { return nil }

  daemon, err := dp.Await(ctx)
  if err != nil {
    return err
  }

  wg, err := daemon.TriggerReloadWithoutCompile(epBPFProgWatchdog)
  if err != nil {
    return err 
  }
  wg.Wait() // FIXME should respect 'ctx' cancellation.
}
```

Since this doesn't provide any API to other modules, it just uses `cell.Invoke` to create a job group and register it's start and stop hooks into the hive's lifecycle. `pkg/hive/job` is essentially `pkg/controller` with lifecycle integrated with Hive plus other features (OneShot, Observable, etc.). We did this as we wanted to be able to inspect not only timer jobs but all background jobs (e.g. one-shot "daemon" functions that are retried on error). It's still lacking a REST API for inspecting the jobs though! We'd like over time to merge `pkg/controller` and `pkg/hive/job` (by implementing `pkg/controller` on top of `job.Timer`), but it's not on top of our TODO list right now.

This could be improved further by moving more things out from `*Daemon`. Using `promise.Promise[*Daemon]` is a sign that something should be fixed (it's there just for transitioning out from `*Daemon`). For example here we could move the `Trigger*` functions out from `*Daemon` into its own API. I think everything but `mtu.Configuration` are already available. Alternative that doesn't change things as much is providing a "bridge" by exposing the trigger functions as their own API, but implementing it in terms of `promise.Promise[*Daemon]` (`cell.Provide(func(dp promise.Promise[*Daemon) Triggers { ... })`, `type Triggers interface { TriggerReloadWithoutCompile(ctx context.Context, reason string) error }`). This is also a way to avoid having to implement everything in `daemon/cmd`. If `Triggers` type is defined elsewhere, then `daemon/cmd` can implement it, but cells that use it can be elsewhere in the codebase. Over time we can then migrate functionality away from `daemon/cmd` into places where they make more sense.

Reply:
Very cool. Thank you Jussi! I'll take a pass. I think there is currently an issue with package deps trying to move things out of pkg cmd since the job cell would refer to the daemon via the promise but the cell deps still needs to be specified on the daemon. Will push my initial crack and we can refine from there.